theseus/embodied/robot/forward_kinematics.py

```diff
@@ -0,0 +1,146 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+from typing import List, Optional
+
+from theseus.labs.lie_functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+
+# TODO: Add support for joints with DOF>1
+
+
+def ForwardKinematicsFactory(robot: Robot, link_names: Optional[List[str]] = None):
+    links: List[Link] = []
+
+    if link_names is None:
+        links = robot.links
+    else:
+        links = [robot.link_map[name] for name in link_names]
+
+    link_ids: List[int] = [link.id for link in links]
+
+    ancestors: List[Link] = []
+    joint_ids: List[int] = []
+    for link in links:
+        ancestors += [anc for anc in link.ancestors]
+        joint_ids += link.angle_ids
+    pose_ids = sorted(list(set([anc.id for anc in ancestors] + link_ids)))
+    joint_ids = sorted(list(set(joint_ids)))
+
+    def _forward_kinematics_helper(angles: torch.Tensor):
+        if angles.ndim != 2 or angles.shape[1] != robot.dof:
+            raise ValueError(
+                f"Joint angles for {robot.name} should be {robot.dof}-D vectors"
+            )
+
+        poses: List[Optional[torch.Tensor]] = [None] * robot.num_links
+        poses[0] = angles.new_zeros(angles.shape[0], 3, 4)
+        poses[0][:, 0, 0] = 1
+        poses[0][:, 1, 1] = 1
+        poses[0][:, 2, 2] = 1
+
+        for id in pose_ids[1:]:
+            curr: Link = robot.links[id]
+            joint: Joint = robot.links[id].parent
+            prev: Link = joint.parent
+            relative_pose = (
+                joint.relative_pose(angles[:, joint.id])
+                if joint.id < robot.dof
+                else joint.relative_pose()
+            )
+            poses[curr.id] = se3.compose(poses[prev.id], relative_pose)
+
+        return poses
+
+    def _forward_kinematics_impl(angles: torch.Tensor):
+        poses = _forward_kinematics_helper(angles)
+        return tuple(poses[id] for id in link_ids)
+
+    def _jforward_kinematics_helper(
+        poses: List[Optional[torch.Tensor]],
+    ) -> torch.Tensor:
+        jposes = poses[0].new_zeros(poses[0].shape[0], 6, robot.dof)
+
+        for id in pose_ids[1:]:
+            link: Link = robot.links[id]
+            joint: Joint = link.parent
+            if joint.id >= robot.dof:
```

Comment:
Is this check possible because joints are assumed to be 1D?

Reply:
Yeah, that's our current assumption.